src/cm/communication_manager.rs

```diff
@@ -56,40 +199,131 @@ impl EventChannel {
-        let mut cm_id = MaybeUninit::<*mut rdma_cm_id>::uninit();
-        let ret;
+    pub fn create_id(&mut self, port_space: PortSpace) -> Result<Arc<CommunicationManager>, String> {
+        let cm = Arc::new(CommunicationManager {
+            cm_id: UnsafeCell::new(NonNull::dangling()), // We'll update this shortly
+        });
 
-        unsafe {
-            ret = rdma_create_id(
+        let mut cm_id_ptr = MaybeUninit::<*mut rdma_cm_id>::uninit();
+        let ret = unsafe {
+            rdma_create_id(
                 self.channel.as_mut(),
-                cm_id.as_mut_ptr(),
-                ctx.context as _,
-                PortSpace::Tcp as u32,
-            );
-        }
+                cm_id_ptr.as_mut_ptr(),
+                Arc::downgrade(&cm).into_raw() as *mut std::ffi::c_void,
+                port_space as u32,
+            )
+        };
 
         if ret < 0 {
             return Err(format!("Failed to create cm_id {ret}"));
         }
 
+        let cm_id = unsafe { NonNull::new(cm_id_ptr.assume_init()).unwrap() };
+
+        // Update the cm_id in the CommunicationManager
         unsafe {
-            cm_id.assume_init();
+            *cm.cm_id.get() = cm_id;
+        }
+
+        Ok(cm)
+    }
+
+    pub fn get_cm_event(&mut self) -> Result<Event, String> {
+        let mut event_ptr = MaybeUninit::<*mut rdma_cm_event>::uninit();
+
+        let ret = unsafe { rdma_get_cm_event(self.channel.as_ptr(), event_ptr.as_mut_ptr()) };
+
+        if ret < 0 {
+            return Err(format!("Failed to get cm event {:?}", io::Error::last_os_error()));
         }
 
-        Ok(CommunicationManager {
-            cm_id: unsafe { NonNull::new(*cm_id.as_mut_ptr()).unwrap_unchecked() },
+        let event = unsafe { NonNull::new(event_ptr.assume_init()).unwrap() };
+
+        let cm_id = unsafe {
+            let raw_cm_id = event.as_ref().id;
+
+            if !raw_cm_id.is_null() {
+                if event.as_ref().event == EventType::ConnectRequest as u32 {
+                    // For connect requests, create a new CommunicationManager
+                    let cm = Arc::new(CommunicationManager {
+                        cm_id: UnsafeCell::new(NonNull::new(raw_cm_id).unwrap()),
+                    });
+
+                    (*raw_cm_id).context = Arc::downgrade(&cm).into_raw() as *mut std::ffi::c_void;
+
+                    Some(cm)
+                } else {
+                    // For other events, return the existing CommunicationManager
+                    let context_ptr = (*raw_cm_id).context as *mut CommunicationManager;
+                    if !context_ptr.is_null() {
+                        // Whenever we consumed the raw pointer of [`Weak`], we should create a new one
+                        // and store in in context for next time of calling.
+                        let cm = Weak::from_raw(context_ptr).upgrade().unwrap();
```

Comment:
Consider using [`loom`](https://github.com/tokio-rs/loom) here for testing upgrade failure